crates/kubelet/src/state/common/resources.rs

```diff
@@ -0,0 +1,127 @@
+        self: Box<Self>,
+        provider_state: SharedState<P::ProviderState>,
+        pod_state: &mut P::PodState,
+        pod: Manifest<Pod>,
+    ) -> Transition<P::PodState> {
+        let pod = pod.latest();
+        debug!(pod = %pod.name(), "Preparing to allocate resources for this pod");
+        let device_plugin_manager = provider_state.read().await.device_plugin_manager();
+
+        // Only check for allocatable resources if a device plugin manager was provided.
+        if let Some(device_plugin_manager) = device_plugin_manager {
+            // Create a map of devices requested by this Pod's containers, keyed by container name
+            let mut container_devices: PodResourceRequests = HashMap::new();
+            for container in pod.all_containers() {
+                if let Some(resources) = container.resources() {
+                    if let Some(requests) = &resources.requests {
+                        let extended_resources: HashMap<String, Quantity> = requests
+                            .clone()
+                            .into_iter()
+                            .filter(|(resource_name, _)| {
+                                util::is_extended_resource_name(resource_name)
+                            })
+                            .collect();
+                        container_devices.insert(container.name().to_string(), extended_resources);
+                    }
+                }
+            }
+            // Do allocate for this Pod
+            if let Err(e) = device_plugin_manager
+                .do_allocate(&pod.pod_uid(), container_devices)
+                .await
+            {
+                error!(error = %e);
+                let next = Error::<P>::new(e.to_string());
+                return Transition::next(self, next);
+            }
+
+            // In Pod, set env vars and set HostPath volumes for each `ContainerAllocateResponse`.
+            // TODO: add support for setting container path mounts, env vars, annotations, and device specs (with permissions) specified in the device plugins' `ContainerAllocateResponse`s.
+            if let Some(container_allocate_responses) =
+                device_plugin_manager.get_pod_allocate_responses(pod.pod_uid())
+            {
+                let mut host_paths: Vec<String> = Vec::new();
+                let mut env_vars: HashMap<String, HashMap<String, String>> = HashMap::new();
+                // Get host paths, env vars, and annotations from allocate responses.
+                container_allocate_responses.iter().for_each(|(c, rs)| {
+                    rs.iter().for_each(|r| {
+                        env_vars.insert(c.clone(), r.envs.clone());
+                        r.mounts
+                            .iter()
+                            .for_each(|m| host_paths.push(m.host_path.clone()))
+                    });
+                });
+                let volumes: HashMap<String, VolumeRef> = host_paths
+                    .iter_mut()
```

Comment:
Nit: This can be `into_iter` so you can remove a clone in the next `map`. Definitely doesn't need to be mutable from what I can see